filters/ELMFilter.cpp

```diff
@@ -122,16 +116,61 @@ void ELMFilter::filter(PointView& view)
 
                 // Otherwise this point is classified as noise, and we proceed
                 // to the next lowest value.
-                view.setField(Dimension::Id::Classification, it->second,
+                view.setField(m_Classdim, it->second,
                               m_class);
                 ++num;
             }
         }
     }
 
+    return num;
+
+}
+
+
+void ELMFilter::filter(PointView& view)
+{
+    BOX2D bounds;
+    calculateBounds(view, bounds);
+
+    size_t cols =
+        static_cast<size_t>(((bounds.maxx - bounds.minx) / m_cell) + 1);
+    size_t rows =
+        static_cast<size_t>(((bounds.maxy - bounds.miny) / m_cell) + 1);
+
+    // Make an initial pass through the input PointView to index elevation
+    // values and PointIds by row and column.
+    std::map<uint32_t, std::multimap<double, PointId>> plus_hash;
+    std::map<uint32_t, std::multimap<double, PointId>> neg_hash;
+    for (PointId id = 0; id < view.size(); ++id)
+    {
+        double x = view.getFieldAs<double>(Dimension::Id::X, id);
+        double y = view.getFieldAs<double>(Dimension::Id::Y, id);
+        double z = view.getFieldAs<double>(m_Zdim, id);
+
+        size_t c = static_cast<size_t>(std::floor(x - bounds.minx) / m_cell);
+        size_t r = static_cast<size_t>(std::floor(y - bounds.miny) / m_cell);
+
+        plus_hash[c * rows + r].emplace(z, id);
```

Comment:
This is a relatively expensive way to do this.  We're just gridding points, so we're using twice as much memory to have a minimum and maximum grid.  Changing the cull function to look through the points from the other direction could eliminate the extra grid.

Reply:
Actually, this whole thing can be made much more lean -- you don't need to store all the point IDs to do this.  Most of the time you can determine at insertion time that a point isn't noise because its distance from some other point is less than the threshold.  Perhaps that could be improved in another PR.

Reply:
You can indeed do this without computing bounds and constructing the grid that way. I've created a new branch that offers a slightly modified approach. It doesn't really get us away from storing PointId's though, and I'm not yet convinced that we can. The problem is that when we consider whether or not a point is noise (either min or max), we need to compare it to the next highest/lowest point, which in turn is compared to the next highest/lowest point. We only ever stop iterating once we've found a point that is sufficiently close to the next in order of elevation. We don't really have enough information for a given cell until we've seen all points contained within that cell.